add document elaborating on how to use TLS and examples

hbagdi opened the pull request.

hbagdi commented:
This fixes #331.

hbagdi commented:
/cc @danehans @jpeach @robscott @Miciah

robscott commented on docs-src/tls.md:
```diff
@@ -0,0 +1,328 @@
+For Gateways, there are two connections involved:
+
+- **Downstream**: This is the connection between the client and the Gateway.
+- **Upstream**: This is the connection between the Gateway and Service. The
+  Service in question can be a core Kubernetes Service or any other network
+  endpoint as well.
```
This is really hard to describe well, I've never found a great way to say it, but maybe something like this?
```suggestion
- **Upstream**: This is the connection between the Gateway and backend resources
   specified by routes. These backend resources will usually be Services.
```

robscott reviewed (commented):
Thanks for taking this on! This is going to be a huge addition to our docs.

hbagdi commented on docs-src/tls.md:
```diff
@@ -0,0 +1,328 @@
+        kind: Secret
+        group: core
+        name: foo-example-com-cert
+      routeOverride:
+        certificate: Deny
```
I considered this but decided against it because of two reasons:
- makes it clear that TLS config is ONLY in Listener
- for anybody trying to learn the TLS section of this API, it helps with keeping all bits in front. I've already remoed route selection section. 

That being said, I can reason either ways and don't feel strongly about it.

hbagdi commented on docs-src/tls.md:
````diff
@@ -0,0 +1,328 @@
+`bar.example.com` domains. The certificate for these domains are specified
+in Gateway.
+
+```yaml
````
That is a fair concern but I deleted the YAML files from the `examples` directory to ensure that I'm not increasing any duplication. I think it is better to include text with examples than having examples in a directory full of YAML files. I, as a user, mostly get lost in that.

Is there any upside of having these in YAML files and then referencing them? I guess one upside is that users can grab the URL and `kubectl apply` it.

robscott replied:
I think the biggest upside is that we'd have a way to automatically verify that our examples weren't stale.

hbagdi replied:
Good point. Let me try to see how to get this to work in a separate PR (and hold this until that). I've no experience with mkdocs.

hbagdi replied:
https://github.com/kubernetes-sigs/service-apis/pull/393

hbagdi replied:
I've updated the PR with YAMLs inside `examples` and using file embedding approach.

#394 fixes a bug and should go in before this.

danehans commented on docs-src/concepts.md:
```diff
@@ -1,9 +1,6 @@
-content of this document started from this [API sketch][api-sketch].
-
-[api-sketch]: https://docs.google.com/document/d/1BxYbDovMwnEqe8lj8JwHo8YxHAt3oC7ezhlFsG_tyag/preview
+This document is a deep dive into the reasoning and design for the API.
```
s/the API/Service APIs/

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+`hostname.match` criteria.
+
+In the following example, a Gateway is configured to serve a TLS certificate
+for any request that is served a Gateway.
```
```suggestion
for any request that is served by the Gateway.
```

danehans replied:
Either drop ` that is served a Gateway` or reword.

danehans commented on docs-src/concepts.md:
```diff
@@ -128,19 +125,10 @@ A typical client/gateway API request flow for a gateway implemented using a reve
-  would be a rare edge case and is not worth supporting at this point. The
-  security model outlined a potential approach to enable this in the future, but
-  there does not seem to be a sufficient reason to work towards that now.
+TLS configuration is tied to Gateway listeners. Additionally, TLS certificates
```
COnsider the following minor change:

s/TLS configuration is tied to Gateway listeners./TLS is configured on Gateway listeners./

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+specific hostname that is set in `hostname.name`. This is the recommended way
+of configuring TLS settings in Service APIs as it does not involve any
+redirection and should suffice in the majority of use-cases.
+Please take a look at the examples below for various possibilities.
```
```suggestion
Please take a look at the examples below for various alternatives.
```

danehans replied:
s/take a look at/review/

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+
+### Route and TLS
+
+If `listener.tls.routeOverride.certificate` is set to `Allow`, TLS certificates
```
```suggestion
If `listeners[].tls.routeOverride.certificate` is set to `Allow`, TLS certificates
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+can be configured at the Route-level. This feature is primarily meant for a
+cluster with a self-service model where Service owners bring their own TLS
+certificates. This feature also mirrors the behavior of TLS as defined in the
+Ingress v1 resource. One should use this feature only when the Gateway Admin
+wishes to delegate TLS to Service or App owners.
```
I don't think "Gateway Admin" and "App owner" are defined anywhere.  Is defining "Gateway Admin" covered by https://github.com/kubernetes-sigs/service-apis/issues/95?  I guess we can come back and update this text if needed depending on what role names are defined in https://github.com/kubernetes-sigs/service-apis/issues/95.

danehans commented on docs-src/concepts.md:
```diff
@@ -128,19 +125,10 @@ A typical client/gateway API request flow for a gateway implemented using a reve
+TLS configuration is tied to Gateway listeners. Additionally, TLS certificates
+can be configured at the Route-level for certain self-service use cases.
+
+Please refer to the [TLS document](tls.md) for a deep dive on TLS.
```
Instead of TLS document, maybe the document should be called TLS Details, etc..

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+
+When using this feature, please note that the TLS certificate to serve is chosen
+before an HTTPRoute is selected. This is because the TLS handshake is completed
+before an HTTP request is sent from the client.
```
This makes me wonder:  Does the same restriction apply if `spec.listeners[].routes.kind` is set to `TLSRoute`?  For HTTPRoutes, the restriction derives directly from the interaction between TLS and HTTP, but TLSRoutes aren't subject to the same limitation.  If this hasn't been decided already, I would suggest that we apply the same restriction to TLSRoutes (and perhaps document it here), even if it isn't strictly necessary, in order to provide a consistent API and clear semantics for implementers.

hbagdi replied:
It does apply to TLSRoute. Is this the right place to document it though? Should it not be in API docs rather?

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+before an HTTPRoute is selected. This is because the TLS handshake is completed
+before an HTTP request is sent from the client.
+
+[TLS Certificate in Route](#tls-certificate-in-route) provides an example
+on how this feature can be used.
```
```suggestion
[TLS Certificate in Route](#tls-certificate-in-route) provides an example
of how this feature can be used.
```
Using "of" feels more correct to me, but this might be a regional difference.

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+
+#### TLS in listener
+
+In this example, Gateway is configured to serve `foo.example.com` and
+`bar.example.com` domains. The certificate for these domains are specified
+in Gateway.
```
```suggestion
In this example, the Gateway is configured to serve the `foo.example.com` and
`bar.example.com` domains. The certificate for these domains is specified
in the Gateway.
```

danehans commented on docs-src/tls.md:
````diff
@@ -0,0 +1,138 @@
+
+```yaml
+listeners:
+- protocol: HTTPS # Other possible value is `TLS`
````
This section is titled "Listener and TLS" but the protocol is `HTTPS` with "# Other possible value is `TLS`. Either rename this section to "Listener and HTTPS" or provide background of the `protocol` field before diving into this example.

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+
+#### Wildcard TLS listeners
+
+In this example, the Gateway is configured with wildcard certificate for
+`*.example.com` and a different certificate for `foo.example.com`.
+Since a specific match takes priority, the Gateway will serve the
+`foo-example-com-cert` for requests to `foo.example.com` and the
+`wildcard-example-com-cert` for all other requests.
```
```suggestion
In this example, the Gateway is configured with a wildcard certificate for
`*.example.com` and a different certificate for `foo.example.com`.
Since a specific match takes priority, the Gateway will serve
`foo-example-com-cert` for requests to `foo.example.com` and
`wildcard-example-com-cert` for all other requests.
```
or
```suggestion
In this example, the Gateway is configured with a wildcard certificate for
`*.example.com` and a different certificate for `foo.example.com`.
Since a specific match takes priority, the Gateway will serve
the certificate in `foo-example-com-cert` for requests to `foo.example.com` and
the certificate in `wildcard-example-com-cert` for all other requests.
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+
+#### TLS Certificate in Route
+
+In this example, Gateway is configured with a default certificate that will be
+served for all hostnames. In addition, `tls.routeOverride.certificate` is set to
+`Allow`, meaning Routes can specify TLS certificates for any domains. Next,
+there are two HTTPRoute resources which define certificates for
+`foo.example.com` and `bar.example.com`.
```
```suggestion
In this example, the Gateway is configured with a default certificate that will be
served for all hostnames. In addition, `tls.routeOverride.certificate` is set to
`Allow`, meaning routes can specify TLS certificates for any domains. Next,
there are two HTTPRoute resources which specify certificates for
`foo.example.com` and `bar.example.com`.
```
I think we should use lower-cased "routes" for the general concept, and only capitalize when referring to a specific kind (such as HTTPRoute).

Miciah commented on examples/tls-cert-in-route.yaml:
```diff
@@ -1,22 +1,16 @@
 metadata:
-  name: my-gateway
-  namespace: default
+  name: cert-in-route--gateway
```
Is the `--` significant, or should that be a single `-`?

danehans commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+### Route and TLS
+
+If `listener.tls.routeOverride.certificate` is set to `Allow`, TLS certificates
+can be configured at the Route-level. This feature is primarily meant for a
```
s/at the Route-level/on Routes that are bound to the Gateway/

danehans replied:
Although someone can infer `routeOverride` will override the gateway listener cert with the route cert, it might be a good idea to explicitly call this out here.

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+
+Please note that in case of `Passthrough` TLS mode, no TLS settings take
+effect as the TLS session from the client is NOT terminated at the Gateway.
+The rest of the document assumes that TLS is being terminated at the Gateway.
```
Perhaps state that "Terminate" is the default.  I noticed that some examples specify `mode: Terminate` and some omit it (but get `Terminate` because it is the default).

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+resource.
+
+Please note that the TLS configuration is related to the Service or backend
+resource and not related to a specific Route resource.
```
```suggestion
resource and not related to a specific route resource.
```

Miciah commented on examples/upstream-tls.yaml:
```diff
@@ -0,0 +1,51 @@
+        type: Prefix
+        value: /
+    forwardTo:
+    - serviceName: my-service1
```
```suggestion
    - serviceName: my-service
```
Typo, I think?

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+## Extensions
+
+Both upstream and downstream TLS configs provide an `options` map to add
+additional TLS settings for implementation specific features.
```
```suggestion
additional TLS settings for implementation-specific features.
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+If `hostname.match` is set to `Exact`, then the TLS settings apply to only the
+specific hostname that is set in `hostname.name`. This is the recommended way
+of configuring TLS settings in Service APIs as it does not involve any
+redirection and should suffice in the majority of use-cases.
```
What does "redirection" refer to here?

hbagdi replied:
Redirection in the sense that TLS configuration is directly defined in the Gateway and doesn't require a Cluster Operator to check all routes for TLS definitions.

Miciah replied:
Ah, as in cross-referencing resources?  And does "This" in "This is the recommended way" refer to setting `tls.routeOverride.certificate: Deny`?  So if I understand correctly, this statement could be rephrased as, "Specifying `tls.routeOverride.certificate: Deny` is recommended because it centralizes TLS configuration within the Gateway specification and should suffice for the majority of use-cases."

hbagdi replied:
Updated.

danehans commented on docs-src/tls.md:
```diff
@@ -0,0 +1,138 @@
+wishes to delegate TLS to Service or App owners.
+
+When using this feature, please note that the TLS certificate to serve is chosen
+before an HTTPRoute is selected. This is because the TLS handshake is completed
```
I'm a bit confused by "When using this feature, please note that the TLS certificate to serve is chosen
before an HTTPRoute is selected." If `listener.tls.routeOverride.certificate` is set to `Allow`, then the cert is configured at the route-level.

danehans replied:
Reading further, I think I have a better understanding of what the above sentence is trying to convey. It might be helpful to introduce the `BackendPolicy` here for clarification.

danehans commented on examples/upstream-tls.yaml:
```diff
@@ -0,0 +1,51 @@
+    kind: Service
+    port: 443
+  tls:
+    clientCertificateRef:
+      name: gateway-client-cert
+      group: core
+      kind: Secret
```
It sounds like `clientCertificateRef` is moving to `Gateway`, so either file an issue to track the need to update this doc or update this PR based on the upcoming changes.

danehans commented:
@hbagdi Great job with this doc. I have a few nits.

bowei commented:
LGTM (holding for others comments)

jpeach commented on docs-src/concepts.md:
```diff
@@ -128,19 +125,10 @@ A typical client/gateway API request flow for a gateway implemented using a reve
-  security model outlined a potential approach to enable this in the future, but
-  there does not seem to be a sufficient reason to work towards that now.
+TLS is configured on Gateway listeners. Additionally, TLS certificates
+can be configured at the Route-level for certain self-service use cases.
```
```suggestion
can be configured on Route objects for certain self-service use cases.
```

jpeach commented on docs-src/tls.md:
```diff
@@ -0,0 +1,144 @@
+In the following example, the Gateway serves the TLS certificate
+defined in `default-cert` Secret resource for all requests.
+Although, the example refers to HTTPS protocol, one can also use the same
+feature for TLS-only protocol alongwith TLSRoutes.
```
```suggestion
feature for TLS-only protocol along with TLSRoutes.
```

robscott commented:
@hbagdi It looks like you've addressed most feedback now, is there anything still left? This generally LGTM.

hbagdi commented:
> @hbagdi It looks like you've addressed most feedback now, is there anything still left? This generally LGTM.

That's my understanding as well. Other than [one comment](https://github.com/kubernetes-sigs/service-apis/pull/391#discussion_r504989542) from @Miciah , I'm seeking an approval so that this PR can get merged.

robscott commented:
Thanks! Will let @danehans, @jpeach, or @Miciah add the final LGTM here.

/approve

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,145 @@
+`hostname.match` criteria.
+
+In the following example, the Gateway serves the TLS certificate
+defined in `default-cert` Secret resource for all requests.
```
```suggestion
defined in the `default-cert` Secret resource for all requests.
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,145 @@
+### Route and TLS
+
+If `listeners[].tls.routeOverride.certificate` is set to `Allow`, TLS certificates
+can be configured on Routes that are bound to the Gateway. This feature is
```
```suggestion
can be configured on routes that are bound to the Gateway. This feature is
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,145 @@
+can be configured on Routes that are bound to the Gateway. This feature is
+primarily meant for a cluster with a self-service model where Application developers
+bring their own TLS certificates. This feature also mirrors the behavior of
+TLS as defined in the Ingress v1 resource. One should use this feature only
+when the Cluster Operator wishes to delegate TLS to Service or Application Developer
+the certificate defined in the Route will override any certificate defined in
+the Gateway
```
Looks like this sentence got corrupted?
```suggestion
TLS as defined in the Ingress v1 resource. One should use this feature only
when the Cluster Operator wishes to delegate TLS configuration to the Application Developer.
With this feature, the certificate defined in the route overrides any certificate defined in
the Gateway.
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,145 @@
+
+The following example shows how upstream TLS can be configured. We have
+omitted downstream TLS configuration for simplicity. As noted before, it
+doesn't matter how downstream TLS is configured for the specific Listener or
+Route.
```
```suggestion
doesn't matter how downstream TLS is configured for the specific listener or
route.
```

Miciah commented on docs-src/concepts.md:
```diff
@@ -128,19 +125,10 @@ A typical client/gateway API request flow for a gateway implemented using a reve
-configure TLS on other resources was considered, ultimately TLS configuration on
-Gateway listeners was deemed sufficient for the following reasons:
+TLS is configured on Gateway listeners. Additionally, TLS certificates
+can be configured on Route objects for certain self-service use cases.
```
```suggestion
can be configured on route objects for certain self-service use cases.
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,145 @@
+suffice for the majority of use-cases. Please take a look at the examples below
+for various alternatives.
+
+### Route and TLS
```
```suggestion
### Routes and TLS
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,145 @@
+
+## Downstream TLS
+
+Downstream TLS settings are configured using Listeners at the Gateway level.
```
```suggestion
Downstream TLS settings are configured using listeners at the Gateway level.
```

Miciah commented on docs-src/tls.md:
```diff
@@ -0,0 +1,145 @@
+
+Downstream TLS settings are configured using Listeners at the Gateway level.
+
+### Listener and TLS 
```
```suggestion
### Listeners and TLS 
```

Miciah commented:
/lgtm
Let's address https://github.com/kubernetes-sigs/service-apis/pull/391/files#r505915278 in a follow-up (I agree, the API godoc can cover this).

danehans commented:
/lgtm
